docs/cypress.md

```diff
@@ -0,0 +1,156 @@
+at the start of each test run.
+
+## Writing Tests
+Mostly this is the same advice as for writing any other Cypress test: the Cypress
+docs are well worth a read if you're not already familiar with Cypress testing, eg.
+https://docs.cypress.io/guides/references/best-practices .
+
+### Getting a Synapse
+The key difference is in starting Synapse instances.  Tests use this plugin via
+`cy.task()` to provide a Synapse instance to log into:
+
+```
+cy.task<SynapseInstance>("synapseStart", "consent").then(result => {
+    synapseId = result.synapseId;
+    synapsePort = result.port;
+});
+```
+
+This returns an object with information about the Synapse instance, including what port
+it was started on and the ID that needs to be passed to shut it down again. It also
+returns the registration shared secret (`registrationSecret`) that can be used to
+register users via the REST API.
+
+Synapse instances should be reasonably cheap to start (you may see the first one take a
+while as it pulls the Docker image), so it's generally expected that tests will start a
+Synapse instance for each test suite, ie. in `before()`, and then tear it down in `after()`.
+
+### Synapse Config Templates
+When a Synapse instance is started, it's given a config generated from one of the config
+templates in `cypress/plugins/synapsedocker/templates`. There are a couple of special files
+in these templates:
+ * `homeserver.yaml`:
+   Template substitution happens in this file. Template variables are:
+   * `REGISTRATION_SECRET`: The secret used to register users via the REST API.
+   * `MACAROON_SECRET_KEY`: Generated each time for security
+   * `FORM_SECRET`: Generated each time for security
+ * `localhost.signing.key`: A signing key is auto-generated and saved to this file.
+   Config templates should not contain a signing key and instead assume that one will exist
+   in this file.
+
+All other files in the template are copied recursively to `/data/`, so the file `foo.html`
+in a template can be referenced in the config as `/data/foo.html`.
+
+### Logging In
+This doesn't quite exist yet. Most tests will just want to start with the client in a 'logged in'
+state, so we should provide an easy way to start a test with element in this state. The
+`registrationSecret` provided when starting a Synapse can be used to create a user (porting
+the code from https://github.com/matrix-org/matrix-react-sdk/blob/develop/test/end-to-end-tests/src/rest/creator.ts#L49).
+We'd then need to log in as this user. Ways of doing this would be:
+
+1. Fill in the login form. This isn't ideal as it's effectively testing the login process in each
+   test, and will just be slower.
+1. Mint an access token using https://matrix-org.github.io/synapse/develop/admin_api/user_admin_api.html#login-as-a-user
+   then inject this into element-web. This would probably be fastest, although also relies on correctly
+   setting up localstorage
```

Comment:
Could we not expose some Lifecycle methods on the window and just call `setLoggedIn` instead of relying on setting up LS?

Reply:
yeah, something along those lines - I think mostly it was more that, one way or another, we end up with code that's used just for the tests, but if we can use a code path in element that's supported anyway, that might be better.